Develop ts_sort_students

sergey-lipikhin opened the pull request.

sergey-lipikhin committed:
Complete solution

YaroslavDzyadiv reviewed (approved):
Great job!

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+export function sortStudents(students: Student[],
+  sortBy: SortType,
+  order: SortOrder): Student[] {
+    type CountAvarageGradeCallback = (grades: number[]) => number;
+    type SortRuleCallback = (
+      a: StudentForSort,
+      b: StudentForSort
+    ) => number;
```
Better keep types on a top level

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+        }
+      ));
+
+    // sort callback
```
If you write comments - it says that your code is too complicated to understand even for you)
Let's get rid of them.

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+    // sort and then get rid of auxiliary fields
+    return studentsToSort
+      .sort(studentsSortRule)
+      .map((
+        { marriedNumber, avarageGrade, ...student },
+      ) => (student));
```
Why do you need to map an array? It's totally unnecessary.

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+    let studentsSortRule: SortRuleCallback | undefined;
+
+    // to provide 'asc' or 'desc' for sorting callback
+    const k = order === 'asc' ? 1 : -1;
```
What is `k`?

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+      / grades.length
+    );
+
+    const studentsToSort: StudentForSort[] = students
+      .map((student) => (
+        {
+          ...student,
+          avarageGrade: countAvarageGrade(student.grades),
+          marriedNumber: Number(student.married),
+        }
+      ));
```
You don't need to have a mapped copy of the array. Could you just sort it without unnecessary modifying?

sergey-lipikhin replied:
In my initial solution I wanted to implement sorting in more optimized way.

When we sort students by avarage grade, on each sort iteration we recalculate objective function (avarage) over and over again. So I suggested to did it once, before sorting. That's why I created copy of `Students` with additional auxiliary field `avarageGrade` once at the beginning to use it while sorting and to avoid time losses.

On the other hand I need to iterate whole Students array once again to remove auxiliary 'avarageGrade' before return.

Also I did it with isMarried (convert boolean to number before sorting). I know that this operation costs nothing, so with my solution we rather lost than gain. But why not :)

And the thing is I can not prove my point with mathematical calculations:) As the objective function becomes more complex, my solution becomes more relevant.

sergii-nosachenko commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,80 @@
+          k * a[sortBy] - k * b[sortBy]
+        );
+        break;
+      default:
```
Let's throw an error and describe the reason. Use template string to display faulty `sortBy` value.

sergey-lipikhin committed:
Refactor sortStudents

1. Replace type declaration to upper scope
2. k -> orderCoefficient
3. Remove comments

sergey-lipikhin committed:
Sort without modifying Student object

sergey-lipikhin committed:
Fix some issues

OleksiiBlezniuk commented on src/sortStudents.ts:
```diff
@@ -1,16 +1,50 @@
+export type SortOrder = 'asc' | 'desc';
+type CountAvarageGradeCallback = (grades: number[]) => number;
+
+const countAvarageGrade: CountAvarageGradeCallback
+= (grades) => (
```
```suggestion
const countAvarageGrade: CountAvarageGradeCallback = (grades) => (
```
